### PR TITLE
WIP: Mobx3

### PR DESCRIPTION
**Pull request by mweststrate**
# MobX 3

MobX 3 will be a short lived major version that will be mainly used to deprecate some old behavior with clear warnings while moving to the new desired behavior in MobX. For example function values will be no longer converted to computed values automatically. Which was in some cases convenient, but mostly confusing. Since changing this behavior would change the semantics without build / compile errors, this transition will be done gradually and completed in MobX 4.

* [x] `observable(plainObject)` will now clone instead of enhance the target object. Use `extendObservable(target, target)` if you still want the old behavior. See #649 
* [x] stop autoconverting argument less functions into computed properties, See #532 
* [x] stop sharing global state by default with other mobx instances, See #621 
* [x] Deprecate whyrun
* [x] Remove not all (not recently) deprecated Mobx 2 methods
* [x] Introduce a better modifier mechanism. Will probably not be introduced in `3.0` but `3.1`
* [x] Add support for ES6 `Map` conversion
* [x]  Process open todo's
* [x] Test typescript 2 latest
* [x] Kill stuff deprecated in MobX 2 & 3, including old modifier mechanism, clean up code paths where needed
* [x] Update docs
* [x] fix: https://github.com/mobxjs/mobx/issues/486
* [x] Revisit exception handling #731 
* [x] Undeprecate whyrun
* [x] Normalize observe callbacks for boxed values
* [x] ~Support Symbol.observable on boxed values~ -> non breaking, so 3.1 #677 
* [x] ~Support Symbol.primitive on boxed values~? -> non breaking, so 3.1 #561
* [ ] Update deps of related packages (utils, react bindings, ...)
* [ ] Update examples

### CODE REVIEW

**Comment by JabX** [comment]
Why is `whyRun` being deprecated already? This is the debugging tool I use when I'm trying to understand why a reaction isn't working as expected. I don't think there's anything that could replace it?

**Comment by mweststrate** [comment]
@JabX basically because I had the impression that nobody is using those :) but if there are more people that use it, it can stay. I'll make a twitter poll soon.

File: CHANGELOG.md [hunk 1]
```diff
@@ -59,6 +128,67 @@ observable({
...
+* Passing a function to `observable` will now create a boxed observable refering to that function
+* Deprecated `whyRun` (as it seems hardly used, please let us know if it should stay)
+* The following deprecated methods have been removed:
+  * `autorunUntil`
+  * `trackTransitions`
+  * `fastArray`
+  * `SimpleEventEmitter`
+  * `ObservableMap.toJs` (use `toJS`)
+  * invoking `observe` and `inject` with plain javascript objects
+=======
```
Comment (mweststrate) [inline CHANGELOG.md]: Remove conflict markers

File: flow-typed/mobx.js [hunk 2]
```diff
@@ -0,0 +1,359 @@
+declare module 'mobx' {
```
Comment (mweststrate) [inline flow-typed/mobx.js]: The mobx3 api changes are not propogated to this file yet

File: CHANGELOG.md (hunk 1 above)
Comment (mweststrate) [inline CHANGELOG.md]: Fixed

File: src/api/action.ts [hunk 3]
```diff
@@ -2,6 +2,45 @@ import {invariant, addHiddenProp} from "../utils/utils";
...
+	<T extends Function>(name: string, fn: T): T;
+
+	// named decorator
+	(customName: string): (target: Object, key: string, baseDescriptor?: PropertyDescriptor) => void;
+
+	// unnamed decorator
+	(target: Object, propertyKey: string, descriptor?: PropertyDescriptor): void;
+
+	// .bound
+	<A1, R, T extends (a1: A1) => R>(fn: T): T;
```
Comment (mweststrate) [inline src/api/action.ts]: Fixed

File: src/api/autorun.ts [hunk 4]
```diff
@@ -103,23 +104,16 @@ export function when(arg1: any, arg2: any, arg3?: any, arg4?: any) {
 	return disposer;
 }
 
-export function autorunUntil(predicate: () => boolean, effect: (r: IReactionPublic) => void, scope?: any): Lambda {
```
Comment (mweststrate) [inline src/api/autorun.ts]: Done

File: src/api/autorun.ts [hunk 5]
```diff
@@ -154,67 +148,60 @@ export function autorunAsync(arg1: any, arg2: any, arg3?: any, arg4?: any) {
...
 }
 
-/**
- *
- * Basically sugar for computed(expr).observe(action(effect))
- * or
- * autorun(() => action(effect)(expr));
- */
-export function reaction<T>(name: string, expression: () => T, effect: (arg: T, r: IReactionPublic) => void, fireImmediately?: boolean, delay?: number, scope?: any): Lambda;
+export interface IReactionOptions {
```
Comment (mweststrate) [inline src/api/autorun.ts]: Done

File: src/api/computed.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+const computedDecorator = createComputedDecorator(false);
+const computedStructDecorator = createComputedDecorator(true);
+
+/**
+ * Decorator for class properties: @computed get value() { return expr; }.
+ * For legacy purposes also invokable as ES5 observable created: `computed(() => expr)`;
+ */
+export var computed: IComputed = (
```
Comment (mweststrate) [inline src/api/computed.ts]: Done

File: src/api/computed.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+		}
+		invariant(typeof arg1 === "function", "First argument to `computed` should be an expression. If using computed as decorator, don't pass it arguments");
+		invariant(arguments.length < 3, "computed takes one or two arguments if used as function");
+		const opts: IComputedValueOptions<any> = typeof arg2 === "object" ? arg2 : {};
+		opts.setter = typeof arg2 === "function" ? arg2 : opts.setter;
+		return new ComputedValue(arg1, opts.context, opts.compareStructural || false, opts.name || arg1.name || "", opts.setter);
+	}
+) as any;
+
+computed.struct = computedStructDecorator;
```
Comment (mweststrate) [inline src/api/computed.ts]: Done

File: src/api/extendobservable.ts [hunk 8]
```diff
@@ -1,34 +1,39 @@
...
- * Extends an object with reactive capabilities.
- * @param target the object to which reactive properties should be added
- * @param properties the properties that should be added and made reactive
- * @returns targer
- */
 export function extendObservable<A extends Object, B extends Object>(target: A, ...properties: B[]): A & B {
+	return extendObservableHelper(target, deepEnhancer, properties) as any;
+}
+
+export function extendShallowObservable<A extends Object, B extends Object>(target: A, ...properties: B[]): A & B {
```
Comment (mweststrate) [inline src/api/extendobservable.ts]: Done

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,16 +1,39 @@
 # 3.0.0
 
+The changelog of MobX 3 might look quite overwhelming, but migrating to MobX 3 should be pretty straight forward nonetheless.
+The api has now become more layered, and the api is more uniform and modifiers are cleaned up.
+In practice, you should check your usage of modifiers (`asFlat`, `asMap` etc.). Besides that the migration should be pretty painless.
+Please report if this isn't the case!
+Note that no chagnes to the runtime algorithm where made, almost all changes evolve in making the creation of observables more uniform, and removing deprecated stuff.
```
Comment (mattiamanzati) [inline CHANGELOG.md]: Typo: changes

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,16 +1,39 @@
...
+
+When passing a plain object to `observable`, MobX used to modify that object inplace and give it observable capabilities.
+This also happened when assigning a plain object to an observable array etc.
+However, this behavior has changed for a few reasons
+
+1.  Both arrays and maps create new data structure, however, `observable(object)` didn't
+2.  It resulted in unnecessary and confusing side effects. If you passed an object you received from some api to a function that added it, for example, to an observable collection. Suddenly your object would be modified as side effect of passing it down to that function. This was often confusing for beginners and could lead to subtle bugs.
+3.  If MobX in the future uses Proxies behind the scenes, this would need to change as well
+
+If you want, you can still enhance existing plainObjects, but simply using `extendObserable(data, data)`. This was actually the old implementation, which has now changed to `extendObservable({}, data)`.
```
Comment (mattiamanzati) [inline CHANGELOG.md]: Typo: "extendObservable" instead of "extendObserable" ;)

**Pull request merged**